xarray/testing/strategies.py

```diff
@@ -0,0 +1,345 @@
+
+    >>> variables(dims=st.just({"a": 2, "b": 1})).example()  # doctest: +SKIP
+    <xarray.Variable (a: 2, b: 1)>
+    array([[-1.00000000e+007+3.40282347e+038j],
+           [-2.75034266e-225+2.22507386e-311j]])
+    """
+
+    if any(
+        not isinstance(arg, st.SearchStrategy) and arg is not None
+        for arg in [dims, dtype, attrs]
+    ):
+        raise TypeError(
+            "Contents dims, dtype, and attrs must each be provided as a hypothesis.strategies.SearchStrategy object (or None)."
+            "To specify fixed contents, use hypothesis.strategies.just()."
+        )
+
+    _array_strategy_fn: ArrayStrategyFn
+    array_strategy: st.SearchStrategy[T_DuckArray]
+    if array_strategy_fn is None:
+        _array_strategy_fn = np_arrays  # type: ignore[assignment]
+    elif not callable(array_strategy_fn):
+        raise TypeError(
+            "array_strategy_fn must be a Callable that accepts the kwargs dtype and shape and returns a hypothesis "
+            "strategy which generates corresponding array-like objects."
+        )
+    else:
+        _array_strategy_fn = (
+            array_strategy_fn  # satisfy mypy that this new variable cannot be None
+        )
+
+    _dtype = draw(dtype)
+
+    if dims is not None:
+        # generate dims first then draw data to match
+        _dims = draw(dims)
+        if isinstance(_dims, Sequence):
+            dim_names = list(_dims)
+            valid_shapes = npst.array_shapes(min_dims=len(_dims), max_dims=len(_dims))
+            _shape = draw(valid_shapes)
+            array_strategy = _array_strategy_fn(shape=_shape, dtype=_dtype)
+        elif isinstance(_dims, (Mapping, dict)):
+            # should be a mapping of form {dim_names: lengths}
+            dim_names, _shape = list(_dims.keys()), tuple(_dims.values())
+            array_strategy = _array_strategy_fn(shape=_shape, dtype=_dtype)
+        else:
+            raise TypeError(
+                f"Invalid type returned by dims strategy - drew an object of type {type(dims)}"
+            )
+
+        _data = draw(array_strategy)
+        if _data.shape != _shape:
+            raise TypeError(
+                "array_strategy_fn returned an array object with a different shape than it was passed."
+                f"Passed {_data.shape}, but returned {_shape}."
+                "Please either specify a consistent shape via the dims kwarg or ensure the array_strategy_fn callable "
+                "obeys the shape argument passed to it."
+            )
+
+    else:
+        # nothing provided, so generate everything consistently
+        # We still generate the shape first here just so that we always pass shape to array_strategy_fn
+        _shape = draw(npst.array_shapes())
+        array_strategy = _array_strategy_fn(shape=_shape, dtype=_dtype)
+        _data = draw(array_strategy)
+        dim_names = draw(dimension_names(min_dims=_data.ndim, max_dims=_data.ndim))
```

Comment:
It's not clear to me if the `shape` and `dtype` args to `array_strategy_fn` should be allowed to be `None`. If they can be `None` then we have to deal with all those cases, but if they can't be `None` then it's on the user to ensure that the `dims` strategy they pass is compatible with the array that gets created, which could be quite confusing.

Reply:
Actually they probably should not be allowed to be `None` to match the behaviour of `hypothesis.extra.numpy.arrays`.

Reply:
I've forbidden them from being `None`.